pkg/logs/input/container/launcher.go

```diff
@@ -18,27 +18,47 @@ import (
 )
 
 // NewLauncher returns a new container launcher depending on the environment.
-// As a user can run on Kubernetes and mount both '/var/log/pods'
-// and the docker socket to collect metrics,
-// we first attempt to initialize the kubernetes launcher
-// and fallback to the docker launcher if the initialization failed.
-func NewLauncher(collectAll bool, sources *config.LogSources, services *service.Services, pipelineProvider pipeline.Provider, registry auditor.Registry) restart.Restartable {
-	// attempt to initialize a kubernetes launcher
-	kubernetesLauncher, err := kubernetes.NewLauncher(sources, services, collectAll)
-	if err == nil {
-		log.Info("Kubernetes launcher initialized")
-		return kubernetesLauncher
-	}
-	log.Infof("Could not setup the kubernetes launcher: %v", err)
+// By default returns a docker launcher if the docker socket is mounted and fallback to
+// a kubernetes launcher if '/var/log/pods' is mounted ; this behaviour is reversed when
+// collectFromFiles is enabled.
+// Returns a noop launcher is none of those volumes are mounted.
+func NewLauncher(collectAll bool, collectFromFiles bool, sources *config.LogSources, services *service.Services, pipelineProvider pipeline.Provider, registry auditor.Registry) restart.Restartable {
+	var (
+		launcher restart.Restartable
+		err      error
+	)
+
+	if collectFromFiles {
```

Comment:
Given that you have early returns you could do:

```
if collectFromFile {
  try k8s
}
try docker
try k8s
```
?

Reply:
I could, it's just a nits right ?

Reply:
As discussed IRL, I feel like it's less readable and I'll stick to code duplication here.